python script enhancements - fix missing Element ID when order

bonidier opened the pull request:
Hi @bittner,

Following remaining part in issue #8, here is the pull request to map missing element ID with a chain of updated ID ... and more !

Regards,
Didier

bonidier committed:
python script enhancements - fix missing Element ID when order
- add CSV file for missing Element ID, to manage a chain of updated ID
- if an element ID is not found, try with updated ID list
- avoid duplicate Element ID
- Selenium:ensure the Element ID return only one result before adding to cart
- rework output

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -90,6 +90,28 @@ def main():
+    load list of new element IDs
+    """
+    global newelementid_map
+    global newelementid_comment
```
I cannot allow code with `global` to slip in. ([Why are global variables evil?](https://stackoverflow.com/questions/19158339/why-are-global-variables-evil#19158418))

Please create a class if you need to share values between functions.

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -339,67 +365,132 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
+        # otherwise the quantity will be set to the previous part
+        if part_no in added_part.keys():
+            duplicate_part_counter += 1
+            print("IGNORE: Already added !".format(pn=part_no))
```
French spacing again ! (Got you!) :smile:

bonidier replied:
Man, "you really got me" :musical_note:

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -339,67 +365,132 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
+                        if part_no in newelementid_comment.keys():
+                            print("\tcomment: {comment}".
+                                  format(comment=newelementid_comment[original_part_no]))
+
```
This empty line seems unnecessary.

bittner commented on lego-mindstorms-pieces.py:
```diff
@@ -339,67 +365,132 @@ def order(shop=None, browser=None, lego_set=None, order_list=None, username=None
+                        sleep(.2)  # seconds
+                        found_counter += 1
+
+                        # quit the part_no chain loop
```
This comment is a (relatively strong) indication that the loop body is significantly too long. Try moving some parts of the loop body into helper functions to make the code easier to understand. When everything is logical, even without comments, then you're good.

bonidier committed:
python code review on order() function:
- Avoid using globals when loading new element ID array
- add querystring to shop URL to fix age survey
- add a try block on age survey
- reduce length of loop about element ID chain

bonidier commented:
Hi @bittner,

To reduce the loop, I'd moved some code outside, but not in a helper : there is too many variables to manage.

It's more readable for now.

The order() function is already long. In my mind it can be better (in a near future) to refactor the content under OOP practice, for maintainability (like using a method per step)

bittner commented:
Wow! The `order` function is now _huge_! Can you split this up a bit?

I think such a huge function reduces readability and thus both maintainability and the probability of future contributions. We need to take care this code is in good shape. Minimum split long function bodies up into several functions, or even move large pieces that belong together in separate classes and/or modules.

I don't want to be the dog in the manger, but I only have two choices: Either oppose and continue maintaining this repo, or let this code creep into the repo and let maintenance go.  :cry:

bonidier commented:
Hi @bittner,
According to our common idea, I'll split the order() function for maintenance purpose

bonidier committed:
[python] maintenance on huge order() function:
- move whole content and linked functions to a module 'legoshop'
- split code into many methods, for readability
- enhancement: read electric part CSV once

bonidier commented:
Hi @bittner,
Here is a first try to move order() content in a module

Regards !

bittner commented:
Not bad! Starts to look nice, doesn't it? -- (**NOTE:** _It will be more complicated for Python beginners, because there is more than a single script to download, but that's the price for getting more features._)

Try to move the helper functions you have (the ones that start with `_` and `__`) out into a helper module or so. And don't forget to run [flake8](http://flake8.pycqa.org/en/latest/index.html#quickstart) over your code! Gives you great feedback.

bonidier committed:
[python] manage LEGO's cookie acceptance
- website has added a banner to accept their cookies
- data: add new mapping for elementid #4666579 (set 45544)

bonidier committed:
[python] fix Lego's website authentication
- Lego has added/chaned some CSS classes

bonidier commented:
Hi Peter, 
I don't forget your feedback about helper module suggestion.
Here is some fixes about Lego's website evolutions since last commit

bonidier committed:
python legoshop module review
- add class to manage updated part id
- add class to manage Mindstorm electric parts
- add class to manage Lego's website access
- pylint pass

bittner commented:
Thanks for your follow-up. Please, give me a few days for reviewing.

Thank you for your understanding.

bonidier commented:
Hi @bittner, 
Thanks for your reactivity, take your time.

Just a note about survey: Lego has changed their survey form, from now we must play with HTML area tag,
I don't now how to fix it, my non-working try is commented.

fortunately, the survey does not appears at each execution

bonidier committed:
[python] bugfix on order when quantity of part is set
- last added part quantity was overrided if current is out-of-stock or electric

bonidier committed:
[python] update elements id mappings
- add new elements id mapping for sets 31313 and 45544
- remove mapping for element id 4514557, lego.com reuse this one

bittner commented:
I'm sorry I obviously forgot to merge these changes! Should I merge them now?

bonidier committed:
[python] update mapping file about ElementID for set 45544 which must be added manually

bonidier commented:
Hi, thanks for your approval!
I've just relaunched order for Home/Education sets for a last review:
Here is a fix for set 45544 about missing ElementIDs to add manually (out of the set).

Now, you can merge
Regards!

bittner commented:
Sorry for the delay. I didn't get the GitHub notifications for this repo for some reason. Strange.

bittner commented:
Thanks for your contribution! :+1:
